File: composeApp/src/androidMain/kotlin/de/carlavoneicken/birthdaysapp/presentation/EditBirthdayScreen.kt
package de.carlavoneicken.birthdaysapp.presentation

import android.widget.Toast
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.OutlinedTextFieldDefaults
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import de.carlavoneicken.birthdaysapp.R
import de.carlavoneicken.birthdaysapp.business.viewmodels.EditBirthdayViewModel
import de.carlavoneicken.birthdaysapp.utils.BackgroundLight
import de.carlavoneicken.birthdaysapp.utils.GoldPrimary
import de.carlavoneicken.birthdaysapp.utils.OrangeAccent
import de.carlavoneicken.birthdaysapp.utils.TextPrimary
import org.jetbrains.compose.ui.tooling.preview.Preview
import org.koin.compose.viewmodel.koinViewModel
import org.koin.core.parameter.parametersOf

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun EditBirthdayScreen(
    birthdayId: Long? = null, // null for new birthday, id for editing
    onDone: () -> Unit = {}
) {

    // context for displaying the success message as a toast
    val context = LocalContext.current
    val viewModel: EditBirthdayViewModel = koinViewModel {
        parametersOf(birthdayId)
    }
    val uiState by viewModel.uiState.collectAsState()

    // LaunchedEffect runs a coroutine whenever the key(s) passed to it change
    // here: "run this block whenever uiState.successMessage changes"
    LaunchedEffect(uiState.successMessage) {
        uiState.successMessage?.let { message ->
            // Show a Toast
            Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
            // Navigate back or do any other side-effect
            onDone()
            // Reset state so it doesn’t re-trigger
            viewModel.clearMessages()
        }
    }

    Scaffold(
        topBar = {
            TopAppBar(
                title = {
                    Text(
                        text = if (uiState.isNew) "New Birthday" else "Edit Birthday",
                        fontSize = 20.sp,
                        fontWeight = FontWeight.Bold,
                        color = TextPrimary
                    )
                },
                navigationIcon = {
                    IconButton(onClick = { onDone() }) {
                        Icon(
                            imageVector = Icons.Default.Close,
                            contentDescription = "Cancel",
                            tint = TextPrimary
                        )
                    }
                },
                colors = TopAppBarDefaults.topAppBarColors(
                    containerColor = BackgroundLight,
                    titleContentColor = TextPrimary
                )
            )
        },
        containerColor = BackgroundLight
    ) { paddingValues ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(paddingValues)
                .verticalScroll(rememberScrollState())
                .padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp)
        ) {
            uiState.errorMessage?.let { error ->
                Card(
                    colors = CardDefaults.cardColors(
                        containerColor = Color(0xFFFFEBEE)
                    ),
                    modifier = Modifier.fillMaxWidth()
                ) {
                    Text(
                        text = error,
                        color = Color(0xFFc62828),
                        modifier = Modifier.padding(16.dp)
                    )
                }
            }

            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                Icon(
                    painter = painterResource(id = R.drawable.ic_person_24),
                    contentDescription = "Person",
                    tint = OrangeAccent
                )
                Text(
                    text = "Name",
                    fontSize = 20.sp,
                    color = OrangeAccent,
                    fontWeight = FontWeight.Bold,
                    modifier = Modifier.padding(start = 10.dp)
                )
            }
            // Name field
            OutlinedTextField(
                value = uiState.name,
                onValueChange = { viewModel.updateName(it) },
                label = { Text("Name") },
                placeholder = { Text("Enter name") },
                singleLine = true,
                modifier = Modifier.fillMaxWidth().padding(bottom = 20.dp),
                colors = OutlinedTextFieldDefaults.colors(
                    focusedBorderColor = OrangeAccent,
                    focusedLabelColor = OrangeAccent,
                    unfocusedContainerColor = Color.White,
                    focusedContainerColor = Color.White
                )
            )

            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                Icon(
                    painter = painterResource(id = R.drawable.ic_calendar_24),
                    contentDescription = "Calendar",
                    tint = OrangeAccent
                )
                Text(
                    text = "Birth Date",
                    fontSize = 20.sp,
                    color = OrangeAccent,
                    fontWeight = FontWeight.Bold,
                    modifier = Modifier.padding(start = 10.dp)
                )
            }

            // Date fields row
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                // Day field
                OutlinedTextField(
                    value = uiState.day,
                    onValueChange = { viewModel.updateDay(it) },
                    label = { Text("Day") },
                    placeholder = { Text("DD") },
                    singleLine = true,
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
                    modifier = Modifier.weight(1f),
                    colors = OutlinedTextFieldDefaults.colors(
                        focusedBorderColor = OrangeAccent,
                        focusedLabelColor = OrangeAccent,
                        unfocusedContainerColor = Color.White,
                        focusedContainerColor = Color.White
                    )
                )

                // Month field
                OutlinedTextField(
                    value = uiState.month,
                    onValueChange = { viewModel.updateMonth(it) },
                    label = { Text("Month") },
                    placeholder = { Text("MM") },
                    singleLine = true,
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
                    modifier = Modifier.weight(1f),
                    colors = OutlinedTextFieldDefaults.colors(
                        focusedBorderColor = OrangeAccent,
                        focusedLabelColor = OrangeAccent,
                        unfocusedContainerColor = Color.White,
                        focusedContainerColor = Color.White
                    )
                )

                // Year field (optional)
                OutlinedTextField(
                    value = uiState.year,
                    onValueChange = {
                        if (it.length <= 4 && it.all { char -> char.isDigit() }) {
                            viewModel.updateYear(it)
                        }
                    },
                    label = { Text("Year") },
                    placeholder = { Text("YYYY") },
                    supportingText = { Text("Optional") },
                    singleLine = true,
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
                    modifier = Modifier.weight(1.2f),
                    colors = OutlinedTextFieldDefaults.colors(
                        focusedBorderColor = OrangeAccent,
                        focusedLabelColor = OrangeAccent,
                        unfocusedContainerColor = Color.White,
                        focusedContainerColor = Color.White
                    )
                )
            }


            Spacer(modifier = Modifier.weight(1f))

            // Action buttons
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                // Cancel button
                OutlinedButton(
                    onClick = { onDone() },
                    modifier = Modifier
                        .weight(1f)
                        .height(56.dp),
                    colors = ButtonDefaults.outlinedButtonColors(
                        contentColor = Color.Black
                    )
                ) {
                    Text(
                        text = "Cancel",
                        fontSize = 18.sp,
                        fontWeight = FontWeight.Medium
                    )
                }

                // Save button
                Button(
                    onClick = {
                        viewModel.saveBirthday()
                    },
                    modifier = Modifier
                        .weight(1f)
                        .height(56.dp),
                    colors = ButtonDefaults.buttonColors(
                        containerColor = GoldPrimary,
                        contentColor = Color.White
                    )
                ) {
                    Text(
                        text = if (uiState.isNew) "Save" else "Update",
                        fontSize = 18.sp,
                        fontWeight = FontWeight.Bold
                    )
                }
            }
        }
    }
}

@Preview
@Composable
fun EditBirthdayScreenPreview() {
    EditBirthdayScreen(birthdayId = 1)
}

File: iosApp/iosApp/Presentation/EditBirthdayScreen.swift
//
//  EditBirthdayScreen.swift
//  iosApp
//
//  Created by Carla von Eicken on 31.10.25.
//

import SwiftUI
import Shared
import KMPObservableViewModelSwiftUI

struct EditBirthdayScreen: View {
    let birthdayId: Int64?
    @StateViewModel var viewModel: EditBirthdayViewModel
    
    @Environment(\.dismiss) private var dismiss
    @EnvironmentObject private var toastCenter: ToastCenter
    
    init(birthdayId: Int64?) {
        self.birthdayId = birthdayId
        
        let kotlinBirthdayId: KotlinLong? = if let id = birthdayId {
            KotlinLong(value: id)
        } else {
            nil
        }
        
        _viewModel = StateViewModel(
            wrappedValue: EditBirthdayViewModel(birthdayId: kotlinBirthdayId)
        )
    }
    
    @FocusState private var focusedField: Field?
    enum Field {
        case name, day, month, year
    }
    
    var body: some View {
        
        VStack(alignment: .leading) {
            HStack {
                Image(systemName: "person")
                    .resizable()
                    .scaledToFit()
                    .frame(width: 18, height: 18)
                    .foregroundColor(Color(orangeAccent))
                
                Text("Name")
                    .font(.system(size: 18))
                    .fontWeight(.semibold)
                    .foregroundStyle(Color(orangeAccent))
            }
            .padding(.bottom, 10)
            
            StyledTextField(
                label: "Name",
                text: Binding(
                    get: { viewModel.uiState.name },
                    set: viewModel.updateName
                ),
                focusedField: $focusedField,
                field: .name,
                prompt: "Name"
            )
            .padding(.bottom, 15)
            
            HStack {
                Image(systemName: "calendar")
                    .resizable()
                    .scaledToFit()
                    .frame(width: 18, height: 18)
                    .foregroundColor(Color(orangeAccent))
                
                Text("Birthdate")
                    .font(.system(size: 18))
                    .fontWeight(.semibold)
                    .foregroundStyle(Color(orangeAccent))
            }
            .padding(.bottom, 25)
            
            HStack {
                StyledTextField(
                    label: "Day",
                    text: Binding(
                        get: { viewModel.uiState.day },
                        set: viewModel.updateDay
                    ),
                    focusedField: $focusedField,
                    field: .day,
                    prompt: "DD"
                )
                StyledTextField(
                    label: "Month",
                    text: Binding(
                        get: { viewModel.uiState.month },
                        set: viewModel.updateMonth
                    ),
                    focusedField: $focusedField,
                    field: .month,
                    prompt: "MM"
                )
                StyledTextField(
                    label: "Year (optional)",
                    text: Binding(
                        get: { viewModel.uiState.year },
                        set: viewModel.updateYear
                    ),
                    focusedField: $focusedField,
                    field: .year,
                    prompt: "YYYY"
                )
            }
            Spacer()
            HStack(alignment: .center, spacing: 20) {
                Button {
                    dismiss()
                } label: {
                    Text("Cancel")
                        .font(.system(size: 18))
                        .fontWeight(.semibold)
                        .foregroundStyle(Color(textPrimary))
                        .padding(10)
                        .padding(.horizontal, 30)
                }
                .buttonStyle(.bordered)
                
                Button {
                    viewModel.saveBirthday()
                } label: {
                    Text(viewModel.uiState.isNew ? "Save" : "Update")
                        .font(.system(size: 18))
                        .fontWeight(.semibold)
                        .foregroundStyle(.white)
                        .padding(10)
                        .padding(.horizontal, 30)
                }
                .buttonStyle(.borderedProminent)
                .tint(Color(goldPrimary))
            }
            .frame(maxWidth: .infinity)
        }
        .padding(25)
        .navigationTitle(birthdayId != nil ? "Edit Birthday" : "New Birthday")
        
        // Check if the uiState for success- or errormessage changes
        .onChange(of: viewModel.uiState.successMessage) { _, message in
            if let message {
                viewModel.clearMessages()
                toastCenter.show(Toast(message: message, width: 260))
                dismiss()
            }
        }
        .onChange(of: viewModel.uiState.errorMessage) { _, message in
            if let message {
                viewModel.clearMessages()
                toastCenter.show(Toast(message: message, width: 260))
            }
        }
    }
}


// MARK: Styles Text Field
struct StyledTextField<Field: Hashable>: View {
    let label: String
    @Binding var text: String
    @FocusState var isFocused: Bool
    @FocusState.Binding var focusedField: Field?
    let field: Field
    let prompt: String
    
    var body: some View {
        // The floating label
        ZStack(alignment: .leading) {
            if(label != "Name") {
                Text(label)
                    .foregroundColor(isFocused ? Color(orangeAccent) : .gray)
                    .font(.system(size: 12))
                    .offset(y: -31)
                    .padding(.horizontal, 4)
                    .background(Color.white)
            }
            TextField(label, text: $text, prompt: Text(prompt))
                .focused($isFocused)
                .padding(12)
                .background(Color.white)
        }
        .overlay(
            RoundedRectangle(cornerRadius: 8)
                .stroke(isFocused ? Color(orangeAccent) : Color.gray.opacity(0.5), lineWidth: 1)
        )
        .padding(.bottom, 20)
    }
}
